File: DynamicProgramming/RodCutting/rod_cutting_no_mem.py
from time import time 

def cut_rod(p, n):
    if n == 0:
        return 0
    q = -1 # sentinel value
    for i in range(1, n + 1):
        idx = i - 1
        q = max(q, p[idx] + cut_rod(p, n - i))
    return q


table = [
    1,  # $1 for 1-inch
    3,  # $3 for 2-inch
    5,  # $4 for 3-inch
    6,  # $6 for 4-inch
    9,  # $9 for 5-inch
    9, # $13 for 6-inch
    12, # $17 for 7-inch
    13, # 8
    16, # 9
    26, # 10
    28, # 11
    30, # 12
    32, # 13
    38, # 14
    39, # 15
    42, # 16
    44, # 17
    48, # 18
    50, # 19
    51, # 20
    52, # 21
    57, # 22
    61, # 23
    65, # 24
    67, # 25
    68, # 26
    80, # 27
]

begin = time()
n = cut_rod(table, 21)
end = time()
print(n)
print('executed in {}s'.format(end - begin))

File: DynamicProgramming/RodCutting/rodcutting_no_mem.rs
use std::cmp::max;
use std::env::args;

struct PriceTable {
    table: Vec<isize>,
}

fn main() {
    let p_table = PriceTable::new();
    let m = args().nth(1).expect("no integer given");
    let m: usize = m.parse().unwrap();
    let n = cut_rod(m, &p_table);
    println!("optimal price is: ${}.00", n);
}

fn cut_rod(n : usize, table: &PriceTable) -> isize {
    /*  We have the recurrence
        r_n = max(p_1, r_1 + r_{n-1}, r_2 + r_{n - 2}, ... , r_{n - 1} + r_1) */
    match n {
        // Let the base case be p_0 = 0
        0 => 0,
        // else
        _  => {
            let mut m = -1000;
            for i in 0..n {
                // Let's compare m
                // and table[i] + p_{n - i}
                // we do i + 1 to compensate for the 0 indexing
                // and the fact that 0..n is inclusive of n.
                m = max(m, table.get(i) + cut_rod(n - (i + 1), table));
            }
            m
        }
    }
}

impl PriceTable {
    fn new() -> PriceTable {
        let v = vec![1, 5, 8, 9, 10, 17, 17, 20, 24, 30];

        PriceTable {
            table: v,
        }
    }

    fn get(&self, n: usize) -> isize {
        self.table[n]
    }
}